### PR TITLE
Add QAT support and pass hooks to QuantizationRecipe.

### PR DESCRIPTION
**Pull request by MartinPavella**
### Summary
Extend `QuantizationRecipe` with QAT support and four ordered pass hooks, then wire them into `QuantizeStage`.

`QuantizationRecipe` gains `is_qat`, `train_fn`, `calibration_inputs_fn`, and four pass-list fields (`pre_prepare_passes`, `post_prepare_passes`, `pre_convert_passes`, `post_convert_passes`). When `is_qat=True`, `QuantizeStage` calls `prepare_qat_pt2e` instead of `prepare_pt2e`, invokes `train_fn` on the prepared model, and skips calibration. When `is_qat=False` (default), PTQ proceeds as before, but callers may now supply a `calibration_inputs_fn` to supply their own calibration data. If omitted, the existing example-inputs fallback is used. The four pass hooks are applied at the appropriate points in both flows and compose cleanly with `_combine_recipes`.

`_combine_recipes` is overhauled to handle all previously dropped fields: the four pass lists are concatenated, `is_qat` must agree across all combined recipes, at most one `train_fn` is allowed, and multiple `calibration_inputs_fn` values are chained into a single factory. The `strict`, `mode`, `pipeline_stages`, and `source_transform_in_place` scalar fields are now also validated for agreement and propagated to the combined recipe. `edge_manager_transform_passes` from `LoweringRecipe` is similarly merged.

### Test Plan
```
pytest -q export/tests/test_export_recipe.py export/tests/test_export_stages.py
```

### CODE REVIEW

File: export/stages.py [hunk 1]
```diff
@@ -423,6 +427,15 @@ def _get_quantizer_for_prepare_pt2e(self, quantizers: List[Any]):
         else:
             raise ValueError("No quantizers detected")
 
+    @staticmethod
+    def _apply_passes(
+        model: "torch.fx.GraphModule",
+        passes: Optional[List[Callable]],
+    ) -> "torch.fx.GraphModule":
+        for pass_fn in passes or []:
+            model = pass_fn(model)
```
Comment (novak-vaclav) [inline export/stages.py]: nit: I'd implement some kind of error handling, or at least logging.

File: export/recipe.py [hunk 2]
```diff
@@ -88,16 +88,50 @@ class QuantizationRecipe:
...
+        post_convert_passes: Optional list of callables applied to the GraphModule after convert_pt2e.
+                             Each callable receives a GraphModule and must return a GraphModule.
     """
 
     quantizers: Optional[List[Quantizer]] = None
     ao_quantization_configs: Optional[List[AOQuantizationConfig]] = None
+    is_qat: bool = False
+    calibration_inputs_fn: Optional[Callable[[], Iterable[tuple]]] = None
+    train_fn: Optional[Callable[["torch.fx.GraphModule"], None]] = None
+    pre_prepare_passes: Optional[
```
Comment (novak-vaclav) [inline export/recipe.py:171]: nit: If possible, I'd forbid `None` for these passes and default to `[]`, allowing you to remove the `if ... is not None` statements.

File: export/tests/test_export_recipe.py
Comment (novak-vaclav) [inline export/tests/test_export_recipe.py:1]: I'm not sure about the licensing policy, but I think NXP should be mentioned here as well. May apply to all the changed files.

File: export/recipe.py [hunk 3]
```diff
@@ -249,60 +283,135 @@ def _combine_recipes(  # noqa: C901
...
+                    all_post_convert_passes.extend(qr.post_convert_passes)
+
+            strict_values.append(recipe.strict)
+            mode_values.append(recipe.mode)
+            pipeline_stages_values.append(
+                tuple(recipe.pipeline_stages) if recipe.pipeline_stages else None
+            )
+            source_transform_in_place_values.append(recipe.source_transform_in_place)
+
             if combined_backend_config is None and recipe.executorch_backend_config:
```
Comment (novak-vaclav) [inline export/recipe.py]: I believe this logic causes the first `executorch_backend_config` that is found to be used, however it's not enforced that they are the same across recipes. It may be desired behavior, if so, please add a comment.

File: export/recipe.py [hunk 4]
```diff
@@ -249,60 +283,135 @@ def _combine_recipes(  # noqa: C901
...
+        _assert_agree("strict", strict_values)
+        _assert_agree("mode", mode_values)
+        _assert_agree("pipeline_stages", pipeline_stages_values)
+        _assert_agree("source_transform_in_place", source_transform_in_place_values)
+
+        # is_qat must agree across all recipes that carry a QuantizationRecipe.
+        _assert_agree("is_qat", is_qat_values)
+        # train_fn must have at most one non-None value across all recipes.
+        non_none_train_fns = [f for f in train_fn_values if f is not None]
+        if len(non_none_train_fns) > 1:
```
Comment (novak-vaclav) [inline export/recipe.py:385]: Also `len(non_none_train_fns)` can't be 0 I think

File: export/recipe.py [hunk 5]
```diff
@@ -249,60 +283,135 @@ def _combine_recipes(  # noqa: C901
...
+                    yield from _fn()
+
+            combined_calib_fn = _combined_calib_fn
+        else:
+            combined_calib_fn = non_none_calib_fns[0] if non_none_calib_fns else None
+
+        # Build combined QuantizationRecipe.
         combined_quantization_recipe = None
-        if all_quantizers or all_ao_quantization_configs:
+        if (
```
Comment (novak-vaclav) [inline export/recipe.py]: nit: I'd add an INFO log message in the `else` branch of this `if` to let the user know the `QuantizationRecipe` is all default, which might be suspicious.

File: export/recipe.py [hunk 6]
```diff
@@ -249,60 +283,135 @@ def _combine_recipes(  # noqa: C901
...
+                pre_prepare_passes=all_pre_prepare_passes or None,
+                post_prepare_passes=all_post_prepare_passes or None,
+                pre_convert_passes=all_pre_convert_passes or None,
+                post_convert_passes=all_post_convert_passes or None,
             )
 
         # Create combined lowering recipe
         combined_lowering_recipe = None
-        if all_partitioners or all_transform_passes:
+        if (
```
Comment (novak-vaclav) [inline export/recipe.py]: nit: Same thing here, I'd add a INFO log message in the `else` branch of this `if` to let the user know the `LoweringRecipe` is all default.

File: export/recipe.py [hunk 7]
```diff
@@ -249,60 +283,135 @@ def _combine_recipes(  # noqa: C901
...
+            all_partitioners
+            or all_edge_transform_passes
+            or all_edge_manager_transform_passes
+        ):
             edge_compile_config = None
```
Comment (novak-vaclav) [inline export/recipe.py]: I know this is not part of your changes, but the handling of `edge_compile_config` is weird - similarily to `executorch_backend_config`, the first value found is used - it may be desired behavior, if so, please add a comment why, if you know.

File: export/tests/test_export_recipe.py [hunk 8]
```diff
@@ -129,3 +136,436 @@ def test_get_recipe_with_kwargs_verification(self) -> None:
...
+            quantization_recipe=QuantizationRecipe(quantizers=[Mock()], is_qat=True),
+        )
+        r2 = ExportRecipe(
+            name="b",
+            quantization_recipe=QuantizationRecipe(quantizers=[Mock()], is_qat=True),
+        )
+        result = ExportRecipe.combine([r1, r2])
+        self.assertTrue(result.quantization_recipe.is_qat)
+
+    def test_two_train_fns_raises(self) -> None:
```
Comment (novak-vaclav) [inline export/tests/test_export_recipe.py:572]: I'd also test that for `is_qat=True` and no `train_fn` it raises.

File: export/stages.py [hunk 9]
```diff
@@ -447,15 +461,53 @@ def run(self, artifact: PipelineArtifact) -> None:
...
+                if recipe.train_fn is None:
+                    raise ValueError("train_fn must be provided when is_qat=True")
+                prepared_model = prepare_qat_pt2e(captured_graph, quantizer)
+
+                # Pass 2: post-prepare passes.
+                prepared_model = self._apply_passes(
+                    prepared_model, recipe.post_prepare_passes
+                )
+
+                recipe.train_fn(prepared_model)
```
Comment (novak-vaclav) [inline export/stages.py:556]: Model should be put into `training` mode and then moved back to `eval` mode.

File: export/stages.py [hunk 10]
```diff
@@ -447,15 +461,53 @@ def run(self, artifact: PipelineArtifact) -> None:
...
+            else:
+                prepared_model = prepare_pt2e(captured_graph, quantizer)
```
Comment (novak-vaclav) [inline export/stages.py:559]: Model should be put into `eval` mode just in case.

File: export/stages.py (hunk 9 above)
Comment (novak-vaclav) [inline export/stages.py:556]: AI pointed out that the `train_fn` is the same for all model methods, which seems too restrictive to me. I'd make the `train_fn` a `Dict[str, Callable]` mapping of method to a train function (similarily to `example_inputs`). If you decide to go with this suggestion, additional validation would also have to be implemented (checking there is a training function for each method...)

File: export/tests/test_export_recipe.py [hunk 11]
```diff
@@ -129,3 +136,436 @@ def test_get_recipe_with_kwargs_verification(self) -> None:
...
+            quantization_recipe=QuantizationRecipe(quantizers=[Mock()], is_qat=False),
+        )
+        with self.assertRaises(ValueError) as cm:
+            ExportRecipe.combine([r1, r2])
+        self.assertIn("is_qat", str(cm.exception))
+
+    def test_agreeing_is_qat_preserved(self) -> None:
+        r1 = ExportRecipe(
+            name="a",
+            quantization_recipe=QuantizationRecipe(quantizers=[Mock()], is_qat=True),
```
Comment (novak-vaclav) [inline export/tests/test_export_recipe.py:563]: nit: Technically `quantizers` field doesn't have to filled, I'd leave it empty. This is present in other tests as well.

File: export/recipe.py
Comment (novak-vaclav) [inline export/recipe.py]: nit: `_combine_recipes` is too long, but it's not flagged because of this lint ignore. Imo it's time to decompose it.

File: export/recipe.py [hunk 12]
```diff
@@ -88,16 +88,50 @@ class QuantizationRecipe:
     """
     Configuration recipe for quantization.
 
-    This class holds the configuration parameters for quantizing a model.
+    This class holds the configuration parameters for quantizing a model, supporting
+    both post-training quantization (PTQ) and quantization-aware training (QAT).
 
     Attributes:
```
Comment (novak-vaclav) [inline export/recipe.py:138]: Just wondering, will there be some documentation update on how to use the recipes with all the newly added options? Or a real world example?

**Review by novak-vaclav — COMMENTED** [review]
A few comments, otherwise very good job 😊👌

**Comment by JakeStevens** [comment]
A couple high level thoughts.

IMO QAT will likely be used in two passes, for prepare/convert essentially. 

as is, we have a train_fn that must operate in-place. This works for more simple set ups, but makes it hard to use something like titan or other training frameworks, which mutate the model for model parallelism etc.

A more framework-friendly API would expose separate operations:

```
  prepared = session.prepare_qat()
  # TorchTitan or another framework trains/preempts/resumes/checkpoints it.
  quantized = session.convert_qat(prepared_or_checkpoint)
```

There is also often an need to keep a prepared model (or converted) without finishing off the full lowering, for server evals.

Similary, for training, we need to support a dynamic shape so that we are not setting batch=1 for training, OR batch=<some huge number> for inference and fixing that at graph capture.

You may want to add `allow_exported_model_train_eval()` after preparation

You may want to call `model.train()` before capture

File: export/recipe.py (hunk 4 above)
Comment (MartinPavella) [reply export/recipe.py:385]: It can be 0 if PTQ recipes are combined.

File: export/recipe.py (hunk 2 above)
Comment (MartinPavella) [reply export/recipe.py:171]: It's perfectly fine the way it is. Plus it's inline with how the `quantizers` were handled before. Plus defaulting to `[]` is bad practice.

File: export/tests/test_export_recipe.py
Comment (MartinPavella) [reply export/tests/test_export_recipe.py:1]: Good point. I believe NXP should be mentioned there too. I will add it.

File: export/tests/test_export_recipe.py (hunk 8 above)
Comment (MartinPavella) [reply export/tests/test_export_recipe.py:572]: `is_qat=True` and no `train_fn` is a problem for the lowering, not for recipe combination. Recipe combination raises when the recipes cannot be combined. So the test you are suggesting doesn't make much sense.
(there is already a test that checks this combination raises during lowering (`test_run_qat_missing_train_fn_raises`))

File: export/stages.py (hunk 9 above)
Comment (MartinPavella) [reply export/stages.py:556]: Good point 👍🏻

File: export/stages.py (hunk 1 above)
Comment (MartinPavella) [reply export/stages.py]: I will add code to print which pass failed if an exception is raised.

File: export/recipe.py (hunk 12 above)
Comment (MartinPavella) [reply export/recipe.py:138]: Most likely yes. We will deal with that once the stages and recipes are resolved thogh.

File: export/stages.py (hunk 9 above)
Comment (MartinPavella) [reply export/stages.py:556]: Will leave this as is for now. If a use case comes up, it can always be added.

File: export/tests/test_export_recipe.py (hunk 11 above)
Comment (MartinPavella) [reply export/tests/test_export_recipe.py:563]: Not important.

**Comment by MartinPavella** [comment]
> A couple high level thoughts.
> 
> IMO QAT will likely be used in two passes, for prepare/convert essentially.
> 
> as is, we have a train_fn that must operate in-place. This works for more simple set ups, but makes it hard to use something like titan or other training frameworks, which mutate the model for model parallelism etc.
> 
> A more framework-friendly API would expose separate operations:
> 
> ```
>   prepared = session.prepare_qat()
>   # TorchTitan or another framework trains/preempts/resumes/checkpoints it.
>   quantized = session.convert_qat(prepared_or_checkpoint)
> ```
> 
> There is also often an need to keep a prepared model (or converted) without finishing off the full lowering, for server evals.
> 
> Similary, for training, we need to support a dynamic shape so that we are not setting batch=1 for training, OR batch= for inference and fixing that at graph capture.
> 
> You may want to add `allow_exported_model_train_eval()` after preparation
> 
> You may want to call `model.train()` before capture

Thanks @JakeStevens for the review, this was a really useful comment.

__What I changed in this PR:__ I added QAT support to `QuantizeStage`. The stage now branches on `is_qat`: it uses `prepare_qat_pt2e`, calls `allow_exported_model_train_eval()` on the prepared graph, moves it to train mode, runs the caller-supplied `train_fn`, moves it back to eval, and then `convert_pt2e`. I also added a `dynamic_batch_size` flag so dimension 0 of the calibration/QAT inputs can vary (marks dim 0 dynamic in the internal `torch.export.export` call), plus an optional `calibration_inputs_fn` for PTQ.

__On the broader request:__ If I understand correctly, you'd like to be able to (1) prepare the model for QAT, (2) run QAT with the ability to checkpoint and resume (potentially distributed across machines e.t.c.) and only then (3) optionally lower the model. I don't think that workflow should live inside a single `QuantizeStage`/callback, because an in-process callback forces the whole training loop to run inside one uninterrupted `export()` call and can't span multiple processes or checkpoints.

I believe it can instead be achieved with two recipes / two `export()` calls:

- __Recipe 1__ contains only the stages up to (but excluding) quantization to prepare the graph. In practice this is often just exporting to the aten graph, so this recipe would probably often not even be necessary (the user can do that in ~1 line).
- The user's own scripts then own the QAT loop: `prepare_qat_pt2e`, train, save/resume checkpoints, ... and finally `convert_pt2e`.
- __Recipe 2__ is any backend's recipe (default stages, or specifically only the post-quant stages). You call `executorch.export` with the resulting quantized `GraphModule` and that recipe. Because the input is a `GraphModule`, the default pipeline already skips the quantization stage and just lowers it.

This approach would allow the user to decide whether or not to call `convert_pt2e` or whether to continue with model lowering. Would this approach work from your side? If so, I think it's mostly a matter of docs + an example in the code, and maybe a couple of quality-of-life helpers rather than new stage machinery.

Either way, I think many backends (including Neutron) can already make good use of the basic in-pipeline QAT support in this PR. Would it be OK to merge this as-is? I'd prefer to keep the docs/example for the checkpointed multi-process QAT workflow for a separate PR.

File: export/stages.py [hunk 13]
```diff
@@ -447,15 +461,53 @@ def run(self, artifact: PipelineArtifact) -> None:
             inputs = example_inputs[method_name][0]
             captured_graph = torch.export.export(model, inputs, strict=True).module()
```
Comment (JakeStevens) [inline export/stages.py]: seems dstill missing dyanmic shape support

**Comment by JakeStevens** [comment]
I still do not see anything about move to train/eval mode, i would expect something ~ to

```
  prepared_model = prepare_qat_pt2e(captured_graph, quantizer)
  prepared_model = move_exported_model_to_train(prepared_model)

  recipe.train_fn(prepared_model)

  prepared_model = move_exported_model_to_eval(prepared_model)
  quantized_model = convert_pt2e(prepared_model)
```


train_fn could do this too but this guards against one that does not

**Comment by JakeStevens** [comment]
agree, it would have to be two sessions

File: export/stages.py (hunk 13 above)
Comment (MartinPavella) [reply export/stages.py]: My apologies, I accidently pushed the changes into our internal repo instead.

Everything is ready here now.

**Review by JakeStevens — APPROVED** [review]

**Review by JakeStevens — CHANGES_REQUESTED** [review]
seems like some recipe tests are failing now

**Comment by MartinPavella** [comment]
> seems like some recipe tests are failing now

Fixed. Also re-based and resolved conflicts with @rascani  's [changes](https://github.com/pytorch/executorch/commit/3bc76d0080c541311bbda9a014995d1533f45cd0).

**Comment by MartinPavella** [comment]
Resolved conflicts with @JakeStevens 's [changes](https://github.com/pytorch/executorch/commit/0c97728dba4ebc1f4ed7ab5245a11e49460069a1)

**Comment by MartinPavella** [comment]
The failing unittest linux jobs are crashing while building `pytorch_tokenizers`. I have seen that on other PRs, so it seems to be unrelated to the changes here.
All relevant tests are passing locally.

**Review by JakeStevens — APPROVED** [review]

**Comment by JakeStevens** [comment]
hmm not sure why cla is failing

File: export/stages.py [hunk 14]
```diff
@@ -487,17 +512,73 @@ def run(self, artifact: PipelineArtifact) -> None:
...
+            # mini-batches of arbitrary size through the prepared graph.
+            export_dynamic_shapes = None
+            if recipe.dynamic_batch_size:
+                from torch.export import Dim
+
+                batch = Dim("batch", min=1)
+                export_dynamic_shapes = tuple(
+                    {0: batch} if isinstance(t, torch.Tensor) else None for t in inputs
+                )
+
```
Comment (JakeStevens) [inline export/stages.py:527]: we shouldmove to train mode here

```
  if recipe.is_qat:
      model.train()
```

in case it is not already

Comment (MartinPavella) [reply export/stages.py:527]: Fixed

**Comment by MartinPavella** [comment]
> hmm not sure why cla is failing

Some legal processes are still ongoing inside NXP regarding the CLA. So we haven't signed it yet.
